zkevm-circuits/src/mpt_circuit.rs

```diff
@@ -289,36 +285,14 @@ impl<F: Field> MPTConfig<F> {
                 (MptCellType::Lookup(Table::Fixed), 3, 3, false),
                 (MptCellType::Lookup(Table::Keccak), 1, 3, false),
                 (MptCellType::Lookup(Table::Exp), 2, 3, false),
-                (MptCellType::MemParentSInput, 1, 3, false),
-                (MptCellType::MemParentSTable, 1, 3, false),
-                (MptCellType::MemParentCInput, 1, 3, false),
-                (MptCellType::MemParentCTable, 1, 3, false),
-                (MptCellType::MemKeySInput, 1, 3, false),
-                (MptCellType::MemKeySTable, 1, 3, false),
-                (MptCellType::MemKeyCInput, 1, 3, false),
-                (MptCellType::MemKeyCTable, 1, 3, false),
-                (MptCellType::MemMainInput, 1, 3, false),
-                (MptCellType::MemMainTable, 1, 3, false),
             ],
             0,
             50,
         );
-
-        let parent_s_table = DynamicLookupTable::from(&state_cm, MptCellType::MemParentSTable);
-        let parent_c_table = DynamicLookupTable::from(&state_cm, MptCellType::MemParentCTable);
-        let key_s_table = DynamicLookupTable::from(&state_cm, MptCellType::MemKeySTable);
-        let key_c_table = DynamicLookupTable::from(&state_cm, MptCellType::MemKeyCTable);
-        let main_table = DynamicLookupTable::from(&state_cm, MptCellType::MemMainTable);
-
         let r = 123456.expr();
         let mut cb = MPTConstraintBuilder::new(5, Some(challenges.clone()), None, r.expr());
         meta.create_gate("MPT", |meta| {
             circuit!([meta, cb], {
-                // Populate lookup tables
-                require!(@KECCAK => <KeccakTable as LookupTable<F>>::advice_columns(&keccak_table).iter().map(|table| a!(table)).collect());
-                require!(@FIXED => fixed_table.iter().map(|table| f!(table)).collect());
-                require!(@MULT => mult_table.iter().map(|table| a!(table)).collect());
```

Comment:
I still like loading the lookup table data like this to be honest. Because this was a general way to load lookup tables into the constraint builder (as expressions, so could be manipulated in any way necessary) and so build_lookups` wouldn't have to take any extra lookup data and it would just work. Now only lookup data stored directly in columns can be loaded in (versus any expressions) and the data needs to be passed in `build_lookups`. Sure, it's a bit of edge case normally, but memory stores do actually do this. So a bit less general with this change.

Reply:
Emmm I fell like this syntax `require!(@KECCAK => ...)` should be designated for dynamic table data cuz it correspond to `require!(... => @KECCAK)` that happens dynamically in the code. 
What about I add a load table API that load fixed tables which will be used in building lookup? Honestly it just become of problem of loading it before or after.

Reply:
Yea I think `load_fixed_table()` at the beginning is good cuz you have some fixed table data ready in addition to thee dyn table data added during configuration, so by the end you just say I want for example all the dynamic or fixed`@KECCAK` lookups to be built at the end.

Reply:
Emmm intuitively i always separate read-only memory and random access memory, and that was what I proposed in the first place. I think having these abstraction is easier to explain to user? And it's a cool point to talk about i guess :) Although you're right when you do `meta.lookup_any` everything is exactly the same, but I suggest to keep some separation?

Reply:
If things always work the same, isn't it easier for users? Because now users have to think what kind of lookups they will be doing and depending on that use a different API. Previously it was always the same, just load the table data using `require!(@MY_TABLE => ...)` and you're good to go. Where that table data comes from, the system doesn't have to care.

Reply:
I see your motivation but I kinda wanna insist on this cuz initially I super disliked these patchwork of hecky solution so I refactored, we should have an _UX that conveys our proof paradigm_. Although I recombined the `add_lookup()` function but internally there's still two path:

**Fixed/Static**

- `=>>` adds lookup with `fixed_path` flag on
    - lrc _must_ be performed because we're targeting one cell
    - reduction are optionally performed 
    - _NO_ `LookupData` saved because we don't need extra info to gen column-based lookup constraint
        *side note: in this way only `local_condition` are applied and `regional_condition` are added during popping region, I argue that doesn't hurt our proof integrity cuz we only need to prove the correctness of static memory read, not occurrence (if the program does static reads at a wrong place, it's okay... cuz that doesn't change execution, basically it's wasting its time but we don't care, we save some efforts dealing with conditions). 
- `build_lookups` calls `build_from_table` with loaded fixed table
   - indicates that we're looking up the _READ-ONLY memory_
   - gen one constraint per (`MY_TYPE`, `my_table`)
 
I can further make an abstraction suggesting the column labeled  `MY_TYPE` is the register designated for READ-ONLY memory checks. Hence we should reflect the static nature of fixed table that gets preloaded into the system, just like how I used `=>>` to cue the user: "your data here is going to one designated column!".

**Dynamic**
- `=>` adds lookup with `fixed_path` flag off
    - leaving the options for user to perform rlc and reduction
    - and extra `to_fixed` flag is also available indicating you don't want your data going to static register but you're still aiming to check static memory
    - store the `LookupData` with configs & condition
 - `build_lookups` calls `build_from_data`
    - iterate each data and recover its regional conditions
    - depending on `to_fixed` flag, lookup to
        - dynamic `TableData` with the table's condition captured at its stored time for pure random access
        - static fixed table

The conclusion is **fixed tables** should be a preloaded part of our system! And I think it should be configured prior to execution, so indeed it should have a separate API.

Reply:
> I see your motivation but I kinda wanna insist on this cuz initially I super disliked these patchwork of hecky solution so I refactored, we should have an _UX that conveys our proof paradigm_.

There was some duplicated code that should have been shared for lookups/lookup tables, but otherwise everything used the exact same path as much as possible? Not sure I understand what was patchy about it. Everything went through he constraint builder and the cell manager with almost no special cases, just some internal improvements necessary to clean things up a bit.

> Although I recombined the `add_lookup()` function but internally there's still two path:
> 
> **Fixed/Static**
> 
>  * `=>>` adds lookup with `fixed_path` flag on
>       
>  * lrc _must_ be performed because we're targeting one cell
>  * reduction are optionally performed
>  * _NO_ `LookupData` saved because we don't need extra info to gen column-based lookup constraint

We do need to save it so we know how many lookups are done per state. In the EVM circuit we set the number of rows required by an opcode by how many lookups are done for each lookup type, so the amount of lookups impacts how high we need to make the state height.

> *side note: in this way only `local_condition` are applied and `regional_condition` are added during popping region, I argue that doesn't hurt our proof integrity cuz we only need to prove the correctness of static memory read, not occurrence (if the program does static reads at a wrong place, it's okay... cuz that doesn't change execution, basically it's wasting its time but we don't care, we save some efforts dealing with conditions).

Not sure I totally understand the " static memory read" things, but the regional condition is applied on the storing of the cached value in the cell (because will overlap with other regions), and the lookup condition itself is indeed applied on the stored value so will be `0` when disabled inside the region.

> * `build_lookups` calls `build_from_table` with loaded fixed table
>       
>  * indicates that we're looking up the _READ-ONLY memory_
>  * gen one constraint per (`MY_TYPE`, `my_table`)

But why would users still have to call `build_lookups`? `build_lookups` is a limitation of the current system, not a feature. Ideally users don't have to call `build_lookups` at all, the system should already know exactly what to do, this way the user also can't forget building all required lookups.

> I can further make an abstraction suggesting the column labeled `MY_TYPE` is the register designated for READ-ONLY memory checks. Hence we should reflect the static nature of fixed table that gets preloaded into the system,

What does READ-ONLY memory have to do with lookups?

For the memory, the table values and the lookup inputs are also RLC encoded and so use exactly the same system, so I don't understand the difference you're making here. Lookups are lookups, you may be doing read-only/dynamic/... or whatever, but they are all normal lookups.

 just like how I used `=>>` to cue the user: "your data here is going to one designated column!".

This one makes sense to me because it impacts performance and it's hard for the system to get smart enough to do this automatically. But the reason to do it is still a limitation of the system and ideally would also get figured out automatically by the system.
> 
> **Dynamic**
> 
> * `=>` adds lookup with `fixed_path` flag off
>       
> * leaving the options for user to perform rlc and reduction
> * and extra `to_fixed` flag is also available indicating you don't want your data going to static register but you're still aiming to check static memory

What does it mean to "check static memory"? I just think about lookups on this level, I don't really care about any other higher level abstractions people may use these lookups for.

Ideally the user doesn't have to decide what it wants to do here, it's too complicated and exposes a lot of complexity. The only reason to to these things is to lower the expression degree, and the constraint builder knows exactly what the max required expression degree is. So the constraint builder could automatically do it for the user or throw an error. But we're not there yet, so some exposing is necessary for advanced users to squeeze out more performance.

> * store the `LookupData` with configs & condition
> 
> * `build_lookups` calls `build_from_data`
>       
> * iterate each data and recover its regional conditions
> * depending on `to_fixed` flag, lookup to      
> * dynamic `TableData` with the table's condition captured at its stored time for pure random access
>  * static fixed table

Yeah for the in-place lookups having the regional conditions separately makes a lot of sense to be able to merge them

But also here I don't understand what "pure random access" has to do with it. It's just a lookup that doesn't use a dedicated column to minimize the number of lookups that need to be done per row, I don't see any other difference with other lookups.

> The conclusion is **fixed tables** should be a preloaded part of our system! And I think it should be configured prior to execution, so indeed it should have a separate API.

I only see two different lookups:
- In-place lookups where lookups are simply generated on the fly as needed. All these lookups are done on the active row.
- Out-of-place lookups that store their inputs in a dedicated column so that the number of lookups/row can be reduced by making use of multiple rows. The added benefit of this approach is also that this automatically reduced the degree of the input (and table expressions) used in the lookups, which can be greatly beneficial for prover performance.

Each of these lookups behaves exactly the same for users so they don't have to care about it for the correctness of their circuit. The only reason to pick one or the other is for performance reasons. Functionality wise they are identical, they are just lookups.